ListView contentHeight should include delayRemove-d items

matthewvogt opened the pull request:
When one or more items are in delayRemove state, the ListView's contentHeight property should include their height. This previously failed if the delayRemove items were at the end of the visibleItems
list.

martinjones commented on src/quick/items/qquickitemview.cpp:
```diff
@@ -1512,7 +1512,7 @@ QQuickItemViewPrivate::~QQuickItemViewPrivate()
 bool QQuickItemViewPrivate::isValid() const
 {
-    return model && model->count() && model->isValid();
+    return model && ((model->count() && model->isValid()) || !visibleItems.isEmpty());
```
This worries me. We use isValid() in many places, some of which will surely assume a valid model if this returns true.

matthewvogt replied:
Yes, but there are no test failures from this change, so how to proceed?

matthewvogt replied:
The alternative is to leave isValid() as-is and modify the isValid() test in setContentSize() and anywhere else required to be this expanded version of isValid().  Which is not a good solution for maintainability, but might be preferred in this case?

martinjones commented:
LGTM

matthewvogt commented:
Updated to add auto test.

martinjones commented on tests/auto/quick/qquicklistview/tst_qquicklistview.cpp:
```diff
@@ -7402,6 +7405,75 @@ void tst_QQuickListView::layoutChange()
+
+    if (useDelayRemove) {
+        QCOMPARE(qRound(listview->contentHeight()), qRound(initialContentHeight));
+        QTest::qWait(200);
```
This is a longer delay than necessary. I'd use a shorter delay in the QML and TRY_COMPARE here.

martinjones commented on tests/auto/quick/qquicklistview/tst_qquicklistview.cpp:
```diff
@@ -7402,6 +7405,75 @@ void tst_QQuickListView::layoutChange()
+
+    listview->setProperty("useDelayRemove", useDelayRemove);
+    QMetaObject::invokeMethod(window->rootObject(), removeFunc.constData());
+    QTest::qWait(100);
```
Use TRY_COMPARE rather than qWait.

matthewvogt replied:
Done.

martinjones commented:
Still LGTM

sletta commented:
Is this change upstreamed?

matthewvogt commented:
No.

mikkoharju commented:
Patch applies on top of qtdeclarative/dev without conflicts, odds are there is not much effort in upstreaming.

matthewvogt commented:
Proposed for upstream: https://codereview.qt-project.org/#/c/114911/

matthewvogt committed:
ListView/GridView contentHeight should include delayRemove-d items

When one or more items are in delayRemove state, the ListView's
contentHeight property should include their height. This previously
failed if the delayRemove items were at the end of the visibleItems
list.  Also applies to GridView.

matthewvogt commented:
Updated to match version merged to upstream.

Can someone with privileges please merge this change?

matthewvogt commented:
Thanks!
